### ISSUE TITLE
Add SysnoMap

### ISSUE THREAD

**Issue by jasonwhite** [wangbj/syscalls#22]
We have `SysnoSet` for creating sets of syscalls, but no way to create a static mapping of syscalls to arbitrary types.

Usage would look something like this:
```rust
use syscalls::{Sysno, SysnoMap};

fn read_callback() {
    println!("read called");
}

fn write_callback() {
    println!("write called");
}

// A static mapping of sysno -> function callback.
static SYSCALL_FUNCTIONS: SysnoMap<fn()> = SysnoMap::new(&[
    (Sysno::read, read_callback),
    (Sysno::write, write_callback),
]);

fn main() {
    // Do a constant-time lookup to call `read_callback`.
    SYSCALL_FUNCTIONS[Sysno::read]();
}
```

This could also be done with a large `match` statement, but that is cumbersome and the compiler may not optimize it very well if the `match` is sparse.

The data type would look something like this:
```rust
pub struct SysnoMap<T> {
    table: [Option<T>; Sysno::table_size()],

    // ...or alternatively, use SysnoSet + MaybeUninit to avoid the Option:
    set: SysnoSet, // For knowing which items are initialized
    table: [MaybeUninit<T>; Sysno::table_size()],
}
```

The API should try to mimic the `HashMap` API as much as possible and have as many methods be `const fn` as possible.

**Comment by nyurik** [comment]
I may need something like this as part of https://github.com/JakWai01/lurk - do you still actively work on this project to merge it?

**Comment by jasonwhite** [comment]
@nyurik I don't actively develop this project much anymore because it is pretty stable, but I will absolutely merge any changes you might have! Implementing `SysnoMap` would be a very welcome change indeed.

**Comment by nyurik** [comment]
@jasonwhite thx, i will take a stab at it then.  I wonder if there are any popular crates already that implement hashset and hashmap based on a fixed-size arrays?  It seems like this should be a common functionality, without the need to reimplement?

**Comment by jasonwhite** [comment]
I'm not aware of any other crates that provide this functionality, but I also haven't searched for it. I don't expect the code for this to be much more complicated than `SysnoSet`. We probably don't need to implement the full interface that `HashMap` has. For example, we can probably do without the complicated Entry API.

**Comment by nyurik** [comment]
There is a somewhat funky https://crates.io/crates/array_map that implements both a set and a map, but it is not ideal.  I implemented the map in #32
